storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala

```diff
@@ -57,66 +57,73 @@ object MergeableStore {
     val (collectedMGetResult, getFailures) = collectWithFailures(mGetResult)
 
     val mPutResultFut: Future[Map[K, Future[Option[V]]]] =
-      collectedMGetResult.map { pairs: Seq[(K, (Option[V], Option[V]))] =>
+      collectedMGetResult.map { pairs: Array[(K, (Option[V], Option[V]))] =>
         val pairMap = pairs.toMap
         val mPutResult: Map[K, Future[Unit]] = store.multiPut(pairMap.mapValues(_._2))
         mPutResult.map { case (k, funit) =>
           (k, funit.map { _ => pairMap(k)._1 })
         }
       }
 
+    val missingFn: K => Future[Option[V]] = { k =>
+      getFailures.flatMap { failures =>
+        Future.exception(failures.getOrElse(k, new MissingValueException[K](k)))
+      }
+    }
+
     /**
-     * Combine original keys with result after put.
+     * Combine original keys with result after put and errors with get.
      */
-    FutureOps.liftFutureValues(keySet, mPutResultFut, getFailures)
+    FutureOps.liftFutureValues(keySet, mPutResultFut, missingFn)
   }
 
   /**
    * Collects keyed futures, partitioning out the failures.
    */
-  private[algebra] def collectWithFailures[K, V](fs:Map[K, Future[V]]): (Future[Seq[(K, V)]], Map[K, Future[Nothing]]) = {
+  private[algebra] def collectWithFailures[K, V](fs: Map[K, Future[V]]): (Future[Array[(K, V)]], Future[Map[K, Throwable]]) = {
```

Comment:
The idea of Either[Throwable, V] to Try[V] is great, indeed Try is basically the same in this context and simpler.

Combining successful and failed results will be inefficient since we need to immediately do multiput for the successful ones and so need them separately. By passing them both back in a single future we have to split them up again immediately after. I do agree that returning a single future conveys the behavior more correctly. It feels like a question of performance and code simplicity vs indication of correct intent. I've implemented the return with single future PTAL.